core/src/main/java/brooklyn/management/internal/EntityManagementUtils.java

```diff
@@ -82,41 +100,90 @@
         return app;
     }
 
-    /** convenience for accessing camp */
-    public static Maybe<CampPlatform> getCampPlatform(ManagementContext mgmt) {
-        return BrooklynServerConfig.getCampPlatform(mgmt);
-    }
-    
     /** as {@link #createApplication(ManagementContext, EntitySpec)} but for a YAML spec */
-    public static <T extends Application> T createUnstarted(ManagementContext mgmt, String yaml) {
-        AssemblyTemplate at = getCampPlatform(mgmt).get().pdp().registerDeploymentPlan( new StringReader(yaml) );
-        return createUnstarted(mgmt, at);
+    public static Application createUnstarted(ManagementContext mgmt, String yaml) {
+        return createUnstarted(mgmt, new StringReader(yaml));
     }
-    
-    /** as {@link #createApplication(ManagementContext, EntitySpec)} but for an assembly template */
+
+    public static Application createUnstarted(ManagementContext mgmt, Reader yaml) {
+        EntitySpec<? extends Application> spec = createEntitySpec(mgmt, yaml);
+        return createUnstarted(mgmt, spec);
+    }
+
+    public static EntitySpec<? extends Application> createEntitySpec(ManagementContext mgmt, String yaml) {
+        return createEntitySpec(mgmt, new StringReader(yaml));
+    }
+
+    public static EntitySpec<? extends Application> createEntitySpec(ManagementContext mgmt, Reader yaml) {
+        BrooklynClassLoadingContext loader = createDefaultLoader(mgmt);
+        return createEntitySpec(mgmt, yaml, loader);
+    }
+
     @SuppressWarnings("unchecked")
-    public static <T extends Application> T createUnstarted(ManagementContext mgmt, AssemblyTemplate at) {
-        CampPlatform camp = getCampPlatform(mgmt).get();
-        AssemblyTemplateInstantiator instantiator;
+    public static EntitySpec<? extends Application> createEntitySpec(ManagementContext mgmt, Reader yaml, BrooklynClassLoadingContext loader) {
+        return (EntitySpec<? extends Application>) createSpec(mgmt, yaml, loader);
+    }
+
+    private static EntitySpec<? extends Application> createEntitySpec(String symbolicName, CampPlatform camp,
+                                                                      DeploymentPlan plan, BrooklynClassLoadingContext loader) {
+        AssemblyTemplate at = registerDeploymentPlan(camp, plan, loader);
+        AssemblyTemplateInstantiator instantiator = newInstantiator(at);
+        if (instantiator instanceof AssemblyTemplateSpecInstantiator) {
+            return (EntitySpec<? extends Application>) ((AssemblyTemplateSpecInstantiator)instantiator).createNestedSpec(at, camp, loader, MutableSet.of(symbolicName));
```

Comment:
Make sure that `createNestedSpec` is called only from `BasicBrooklynCatalog` code (as the `createEntitySpec` wrapping it). Atm it's used for creating apps, thus the error.
Also rename the methods coming from `BasicBrooklynCatalog` to indicate that they are used for catalog items.